File: Sources/TintCore/Palettes/PaletteExtractor.swift
public struct ExtractOptions: Sendable {
    /// How many colours to extract.
    public var count = 16

    /// The image is shrunk so its longer side is at most this many pixels
    /// before clustering. 256 keeps a 5K wallpaper to ~65k pixels — fast, and
    /// the palette barely changes compared with using every pixel.
    public var maxSide = 256

    /// Fixed seed: the same image always produces the same palette.
    public var seed: UInt64 = 7

    public init() {}
}

public enum PaletteError: Error, CustomStringConvertible {
    case empty
    case unreadable(String)

    public var description: String {
        switch self {
        case .empty: "the image has no pixels"
        case .unreadable(let why): why
        }
    }
}

public enum PaletteExtractor {
    /// Clusters already-decoded pixels (at most `maxSide` on the longer side).
    public static func extract(pixels: [Rgb], options: ExtractOptions = ExtractOptions()) throws -> Palette {
        guard !pixels.isEmpty else { throw PaletteError.empty }
        let points = pixels.map(Lab.init(rgb:))
        let clusters = KMeans.run(points, k: options.count, seed: options.seed)
        let total = Double(points.count)
        let swatches = clusters
            .sorted { $0.count > $1.count }
            .map { Swatch(color: $0.center.rgb, share: Double($0.count) / total) }
        return Palette(swatches: swatches)
    }

    /// Reads an image file and extracts its palette.
    public static func extract(file path: String, options: ExtractOptions = ExtractOptions()) throws -> Palette {
        let pixels = try ImageLoader.pixels(path: path, maxSide: options.maxSide)
        return try extract(pixels: pixels, options: options)
    }
}

File: src/Tint.Core/Palettes/PaletteExtractor.cs
using SixLabors.ImageSharp;
using SixLabors.ImageSharp.PixelFormats;
using SixLabors.ImageSharp.Processing;
using Tint.Core.Colors;
using Tint.Core.Imaging;

namespace Tint.Core.Palettes;

public sealed record ExtractOptions
{
    /// <summary>How many colours to extract.</summary>
    public int Count { get; init; } = 16;

    /// <summary>
    /// The image is shrunk so its longer side is at most this many pixels before
    /// clustering. 256 keeps a 5K wallpaper to ~65k pixels — fast, and the
    /// palette barely changes compared with using every pixel.
    /// </summary>
    public int MaxSide { get; init; } = 256;

    /// <summary>Fixed seed: the same image always produces the same palette.</summary>
    public int Seed { get; init; } = 7;
}

public static class PaletteExtractor
{
    public static Palette FromFile(string path, ExtractOptions? options = null)
    {
        using Image<Rgb24> image = ImageLoader.Load(path);
        return FromImage(image, options);
    }

    public static Palette FromImage(Image<Rgb24> image, ExtractOptions? options = null)
    {
        options ??= new ExtractOptions();
        if (options.Count < 1)
        {
            throw new ArgumentOutOfRangeException(nameof(options), "Count must be at least 1.");
        }

        using Image<Rgb24> small = image.Clone(ctx =>
        {
            if (image.Width > options.MaxSide || image.Height > options.MaxSide)
            {
                ctx.Resize(new ResizeOptions
                {
                    Size = new Size(options.MaxSide, options.MaxSide),
                    Mode = ResizeMode.Max,
                    // Box averages whole blocks of pixels: no sharpening halos,
                    // which would otherwise invent colours that are not there.
                    Sampler = KnownResamplers.Box,
                });
            }
        });

        var points = new Lab[small.Width * small.Height];
        small.ProcessPixelRows(rows =>
        {
            for (int y = 0; y < rows.Height; y++)
            {
                Span<Rgb24> row = rows.GetRowSpan(y);
                for (int x = 0; x < row.Length; x++)
                {
                    Rgb24 p = row[x];
                    points[(y * row.Length) + x] = Lab.FromRgb(new Rgb(p.R, p.G, p.B));
                }
            }
        });

        KMeans.Cluster[] clusters = KMeans.Run(points, options.Count, options.Seed);
        double total = points.Length;

        Swatch[] swatches =
        [
            .. clusters
                .OrderByDescending(c => c.Count)
                .Select(c => new Swatch(c.Center.ToRgb(), c.Count / total)),
        ];

        return new Palette(swatches);
    }
}
